RecipeMatch/View Controllers/SearchViewController.m

```diff
@@ -0,0 +1,173 @@
+//  RecipeMatch
+//
+//  Created by ellabuysse on 7/25/22.
+//
+
+#import "SearchViewController.h"
+#import "SearchCollectionReusableView.h"
+#import "UIImageView+AFNetworking.h"
+#import "GridRecipeCell.h"
+#import "APIManager.h"
+#import "DetailsViewController.h"
+#import "DraggableView.h"
+#import "EmptyCollectionReusableView.h"
+
+@interface SearchViewController () <UICollectionViewDelegate, UICollectionViewDataSource, UICollectionViewDelegateFlowLayout, UISearchBarDelegate>
+@property (weak, nonatomic) IBOutlet UICollectionView *collectionView;
+@property (weak, nonatomic) IBOutlet UICollectionViewFlowLayout *flowLayout;
+@property (nonatomic, strong) NSMutableArray *recipes;
+@property (nonatomic, strong) NSString *searchText;
+@property (strong, nonatomic) NSTimer * searchTimer;
+@property (strong, nonatomic) UISearchController *searchController;
+@property (strong, nonatomic) SearchCollectionReusableView *searchView;
+@end
+
+@implementation SearchViewController 
+static const float MIN_LINE_SPACING = 10;
+static const float HEIGHT_FACTOR = 1.2;
+static const float MARGIN_SIZE = 7;
+static const float HEADER_SIZE = 50;
+
+- (void)viewDidLoad {
+    [super viewDidLoad];
+    self.collectionView.delegate = self;
+    self.collectionView.dataSource = self;
+    [self.collectionView registerClass:EmptyCollectionReusableView.self forSupplementaryViewOfKind:UICollectionElementKindSectionHeader withReuseIdentifier:@"EmptyView"];
+}
+
+- (void)searchBar:(UISearchBar *)searchBar textDidChange:(NSString *)searchText {
+    self.searchText = searchText;
+    
+    // if a timer is already active, prevent it from firing
+    if (self.searchTimer != nil) {
+        [self.searchTimer invalidate];
+        self.searchTimer = nil;
+    }
+
+    // reschedule the search: in 0.5 second, call the reloadSearch: method on the new textfield content
+    self.searchTimer = [NSTimer scheduledTimerWithTimeInterval: 1
+                                target: self
+                                selector: @selector(reloadSearch:)
+                                userInfo: searchText
+                                repeats: NO];
+}
+
+// dismisses keyboard when search button is clicked
+- (void)searchBarSearchButtonClicked:(UISearchBar *)searchBar {
+    [searchBar resignFirstResponder];
+}
+
+// called when user stops typing to get recipes
+- (void)reloadSearch:(NSTimer *)timer {
+    NSString *query = timer.userInfo;    // strong reference
+    [self getRecipes:^(NSMutableArray *recipes, NSError *error){
+        if(recipes && [self.searchText isEqualToString:query]){ // check that the returning call is for the correct current query
+            self.recipes = recipes;
+            // only reload section 1 to prevent search bar from losing first responder status
+            [self.collectionView reloadSections:[NSIndexSet indexSetWithIndex:1]];
+            
+        } else {
+            //TODO: add failure support
+        }
+    }];
```

Comment:
this will work!
but i think this is not the most optimal way, as you will still be sending requests even when not need them.
so if someone types very slowly, in a rate of `>timer_time` then, you'll be sending all the requests and waiting to get then back.
`NSURLSessionDataTask` supports canceling requests, so you can probably return this object back (from the api manager), and then cancel it, when creating new request here